project_W/runner_manager.py

```diff
@@ -0,0 +1,368 @@
+            "Runner already has an assigned job!"
+        self.assigned_jobs[job.id] = runner
+        runner.assigned_job_id = job.id
+        logger.info(f"Assigned job {job.id} to runner {runner.runner.id}!")
+
+    @synchronized("mtx")
+    def retrieve_job(self, online_runner: OnlineRunner) -> Optional[Job]:
+        """
+        For a given online runner, retrieves the job that it has been assigned.
+        Additionally, if the runner wasn't marked as processing the job yet, it
+        marks it as such. If the runner has not been assigned a job, it returns
+        None and does nothing.
+        """
+        if online_runner.assigned_job_id is None:
+            return None
+        if online_runner.in_process_job is None:
+            online_runner.in_process_job = InProcessJob(
+                runner=online_runner.runner,
+                job_id=online_runner.assigned_job_id
+            )
+        return online_runner.assigned_job()
+
+    @synchronized("mtx")
+    def submit_job_result(self, online_runner: OnlineRunner, result: str, error: bool) -> Optional[str]:
+        """
+        Handles the submission of a job result by a runner. If the runner is not currently
+        processing a job, returns an error message. Otherwise, marks the job as completed/failed
+        by setting either the transcript or the error_msg field of the job, marks the runner as
+        available and returns None.
+        """
+        if online_runner.in_process_job is None:
+            return "Runner is not processing a job!"
+        job = online_runner.in_process_job.job()
+        if error:
+            job.set_error(result)
+        else:
+            job.set_transcript(result)
+        del self.assigned_jobs[job.id]
+        online_runner.in_process_job = None
+        online_runner.assigned_job_id = None
+        logger.info(f"Marked runner {online_runner.runner.id} as available!")
+        return None
+
+    @synchronized("mtx")
+    def enqueue_job(self, job: Job):
+        if job.id in self.job_queue:
+            return False
+        if (runner := self.find_available_runner(job)) is not None:
+            self.assign_job_to_runner(job, runner)
+            return
+        # Append the job to the end of the queue by inserting
+        # it into the job_queue dict with a dummy value.
+        # TODO: Insert using job priority once added.
+        self.job_queue.push(job.id, 0)
+        logger.info(f"No runner available for job {job.id}, enqueuing...")
+
+    @synchronized("mtx")
+    def heartbeat(self, runner: Optional[Runner], req: Request) -> HeartbeatResponse:
+        # TODO: actually handle the request data for job updates and such.
+        if runner is None:
+            return HeartbeatResponse(error="No runner with that token exists!")
+        if not self.is_runner_online(runner):
+            return HeartbeatResponse(error="This runner is not currently registered as online!")
```

Comment:
One question: What if the backend goes briefly offline while a runner is still running? For example let's say a runner is working on a large job, and the administrator decides to update the docker containers (docker compose pull, docker compose up -d) which would restart the server. Then the /api/runners/heartbeat route would return this error message and the job will be queued again in load_jobs_from_db thus removing all current progress, right? 
To be clear this is not a huge problem right now IMO, but maybe we want to keep this in mind for the future...

Reply:
Yes, if the backend goes offline while the runner is processing, the progress of the runner will be discarded. I don't think this is a big issue, because the whole runner scheduling is all session based anyways (i.e. it all resets when the backend restarts), and the backend isn't expected to have regular restarts anyways.